### PR TITLE
[dynamic-shapes] initial support for dynamic shape typechecks

### PR DESCRIPTION
**Pull request by mattjj**
After this PR, jaxpr typechecking works with all existing dynamic shape tests. I was working on dynamic shape AD and realized I could make faster progress with some reliable jaxpr checks!

Summary of changes:
* Changed the logic in `_check_jaxpr` so as to check that type annotations are well-formed via the new `check_type` function (e.g. variable occurrences in types are in scope).
* Changed the logic in `_check_jaxpr` to handle dynamic shapes in inputs and outputs of primitive applications. For the input side, we change the signature of some primitive application typecheck rules to take input atoms rather than just types (see also last bullet in this list). That way we can check things like e.g. an `xla_call` application with a `call_jaxpr` with lambda binders like `n:i32[] x:f32[n]` against arguments like `3:i32[] y:f32[3]` (correct!) or `3:i32[] y:f32[4]` (error!) or `m:i32[] y:f32[k]` (error!). For the output side, we handle primitive application typechecks returning lists of abstract values in which the shapes may contain `InDBIdx` / `OutDBIdx` axis sizes (representing e.g. dependent tuples), in which case to check against the corresponding let-binder types (which have variables in them) we perform a substitution using `substitute_vars_in_output_ty`.
* Changed the signature of `_check_jaxpr` not to take as argument avals to which the jaxpr is to be applied; we check the called jaxpr's lambda-binder types are consistent with the arguments in the rule `_check_call` rather than in the recursive call to `_check_jaxpr`, which itself now just checks a jaxpr for internal consistency.
* Changed the type signature of custom typechecks to support checking calls and broadcasts, basically to something like `Primitive -> [Atom] -> Params -> OutType` where 
```
OutType = [AvalsWithInOutDBIdx]
AvalsWithInOutDBIdx = ShapedArray [AxisSizeOrDBIdx] DType
AxisSizeOrDBIdx = Lit Int | InDBIdx Int | OutDBIdx Int
```

Minor stuff:
* Moved `DBIdx`, `InDBIdx`, and `OutDBIdx` from partial_eval.py to core.py.
* Some typecheck rules returned `None` as the output type, in which case checking the type of the primitive application against the corresponding let-binders would just be skipped. This PR removes that, and adapts the typecheck rules which were returning `None` to return their correct output types.

### CODE REVIEW

File: jax/core.py [hunk 1]
```diff
@@ -2243,51 +2283,73 @@ def ctx_factory():
...
+      if not typecheck(x.aval, x.val):
         ctx, _ = ctx_factory()
-        raise JaxprTypeError(f"Variable '{pp_var(v, ctx)}' not defined")
-      return env[v]
+        raise JaxprTypeError(
+            f"Literal value {x.val} does not match its type annotation "
+            f"{pp_aval(x.aval, ctx)}")
+      return x
+    else:
+      assert False, "syntactically invalid jaxpr"
```
Comment (froystig) [inline jax/core.py:2309]:
```suggestion
      raise JaxprTypeError(f"syntactically invalid jaxpr: attempt to read {x}")
```

File: jax/core.py [hunk 2]
```diff
@@ -2316,27 +2426,41 @@ def check_eqn(prim, in_avals, params):
     out_avals = [out_avals]
   return out_avals, effects
 
-def check_call(ctx_factory, prim, in_avals, params):
+def _check_call(ctx_factory, prim, in_atoms, params):
```
Comment (froystig) [inline jax/core.py:2429]: Might be a nit, but why did this pick up a leading underscore?

**Review by froystig — APPROVED** [review]
Nice work!

File: jax/core.py (hunk 1 above)
Comment (mattjj) [reply jax/core.py:2309]: I wanted to distinguish jaxpr syntax errors from jaxpr type errors, and assume `check_jaxpr` is only given syntactically valid jaxprs (hence the assert). Same below. WDYT?

File: jax/core.py (hunk 2 above)
Comment (mattjj) [reply jax/core.py:2429]: I wanted to make sure there were no existing callers so I changed the name, and also if the function is just a helper not to be used outside this module having an underscore is a nice hint. WDYT?

File: jax/core.py (hunk 1 above)
Comment (mattjj) [reply jax/core.py:2309]: Put another way, we can and should have error checks for these in the `__init__` methods of the relevant classes (e.g. for this line the corresponding check would be in `JaxprEqn.__init__` asserting that each element of `invars` is an `Atom`).

Comment (froystig) [reply jax/core.py:2309]: I see. Yes, I agree it makes sense to distinguish.

With that in mind, should we check for syntactic validity at any point as well? If so, would we do syntax checks under `check_jaxpr`, and raise `JaxprSyntaxError` rather than `JaxprTypeError`?

This question of whether to syntax-check, and where, is broader in scope than this PR, so we can proceed with `assert` for now and return to it.

File: jax/core.py (hunk 2 above)
Comment (froystig) [reply jax/core.py:2429]: No strong opinion either way – just making sure I wasn't missing some new meaning. In this case, I'd update `check_map` to `_check_map` as well to be consistent.

Comment (froystig) [reply jax/core.py:2429]: By the way, if our current `core` were under `_src` (maybe with select symbol exports via `jax.core`), I would slightly prefer no leading underscores.

File: jax/core.py (hunk 1 above)
Comment (mattjj) [reply jax/core.py:2309]: Well, if we have the checks in the `__init__` methods, then it won't be possible to construct a syntactically invalid jaxpr (without mutation), so we may not need a syntax checker.

Comment (froystig) [reply jax/core.py:2309]: Sure, that'd work.

File: jax/core.py (hunk 2 above)
Comment (mattjj) [reply jax/core.py:2429]: Done.
